Add graph optimization 0/x -> 0

f0k opened the pull request:
Resolves #4213.

This is a simple first attempt. It fixes the problems I encountered in #3758, but based on your warnings, it will probably break other things. Let's see! (I didn't run the full test suite yet.)

f0k commented:
FWIW, the Lasagne test suite still passes with this.

abergeron commented:
If we want this, I'm ok with the current code, but don't merge before the release.

nouiz commented on theano/tensor/opt.py:
```diff
@@ -5271,6 +5271,16 @@ def local_intdiv_by_one(node):
+def local_zero_div(node):
+    """0 / x -> 0
+    """
+    if node.op in [T.int_div, T.true_div, T.floor_div]:
```
That isn't good. There is some corner cases when the op is another instance. use:
isinstance(node.op, T.Elemwise) and isinstance(node.op.scalar_op, (theano.scalar.IntDiv, ...))

instead.

nouiz commented on theano/tensor/opt.py:
```diff
@@ -5271,6 +5271,16 @@ def local_intdiv_by_one(node):
+    """0 / x -> 0
+    """
+    if node.op in [T.int_div, T.true_div, T.floor_div]:
+        if local_mul_canonizer.get_constant(node.inputs[0]) == 0:
```
add the parameter `only_process_constants=True`.

f0k replied:
`local_mul_canonizer.get_constant` is a wrapper that does the correct thing already (I copied it from one of the other optimizers, not sure why they don't call `get_scalar_constant_value` directly),

nouiz commented on theano/tensor/tests/test_opt.py:
```diff
@@ -6123,6 +6123,27 @@ def test3(self):
+    mode = theano.compile.mode.get_default_mode().including("local_zero_div")
+    for t in (T.scalar, T.ivector, T.ftensor4):
+        x = t('x')
+        for op in (T.int_div, T.true_div, T.floor_div):
```
floor_div is int_div, so only use one everywhere
scalar/basic.py have 

floor_div = int_div

nouiz commented on theano/tensor/tests/test_opt.py:
```diff
@@ -6123,6 +6123,27 @@ def test3(self):
+            assert output.ndim == y.ndim
+            assert output.type == y.type
+            # and the output should be zero
+            assert theano.tensor.get_scalar_constant_value(output) == 0
```
That is good, but personaly I prefer to compile a full theano function in test and compile it and check its state. I also like to execute the function to make sure the full pipeline work. But all seem good. I'm ok with this test.

f0k replied:
I figured that for testing graph optimizations, it should be enough to just run the optimizer and skip the compilation. This could potentially be applied to some of the other tests to make them faster.

f0k commented:
Addressed @nouiz's comments and rebased onto latest master so one can test locally without the `-rpath` issues. Ready to merge from my side (but feel free to wait until after the release).

lamblin commented:
The release is over, so I'm merging.
